File: advent_of_code/2022/day_06/solution.rb
require_relative "../utils/ruby_utils/constants"


def raw_data(variation)
  File.read("inputs/06_#{variation}.txt").strip
end

def sliding_window(data, n)
  position = n
  data.chars.each_cons(n) do |marker|
    break if marker == marker.uniq
    position += 1
  end

  position
end

def easy(data)
  sliding_window(data, 4)
end

def hard(data)
  sliding_window(data, 14)
end

if __FILE__ == $0
  data = raw_data(INPUT)
  puts easy(data)
  puts hard(data)
end

File: advent_of_code/2022/day_06/solution.py
from utils.py_utils.constants import INPUT


def get_data(variation):
    with open(f"inputs/06_{variation}.txt") as f:
        data = f.read()
    return data


def sliding_window(data, n):
    for i in range(len(data)-n):
        marker = data[i:i+n]

        if len(set(marker)) == n:
            return i + n


def easy(data):
    return sliding_window(data, 4)


def hard(data):
    return sliding_window(data, 14)


if __name__ == "__main__":
    data = get_data(INPUT)
    print(easy(data))
    print(hard(data))
